src/v2/Artsy/Relay/middleware/cache/Cache.ts

```diff
@@ -0,0 +1,141 @@
+import { QueryResponseCache } from "relay-runtime"
+import createLogger from "v2/Utils/logger"
+import RelayQueryResponseCache from "relay-runtime/lib/network/RelayQueryResponseCache"
+import { isDevelopment, isServer } from "lib/environment"
+import { once } from "lodash"
+
+const logger = createLogger("v2/Artsy/middleware/cache/Cache")
+
+export interface CacheConfig {
+  size: number
+  ttl: number
+  disableServerSideCache?: boolean
+}
+
+export class Cache {
+  cacheConfig: CacheConfig
+  enableServerSideCaching: boolean
+  relayCache: RelayQueryResponseCache
+
+  redisCache: {
+    del: (key: string) => Promise<void>
+    expire: (key: string, ttl: number) => Promise<void>
+    flushall: () => Promise<void>
+    get: (key: string) => Promise<string>
+    set: (key: string, value: string) => Promise<void>
+  }
+
+  constructor(cacheConfig: CacheConfig) {
+    this.cacheConfig = cacheConfig
+    this.enableServerSideCaching =
+      isServer && !this.cacheConfig.disableServerSideCache
+    this.initRelayCache()
+    this.initRedisCache()
+  }
+
+  initRelayCache() {
+    this.relayCache = new QueryResponseCache(this.cacheConfig)
+  }
+
+  initRedisCache() {
+    if (!this.enableServerSideCaching) {
+      return
+    }
+
+    const { promisify } = require("util")
+    const redis = require("redis")
+
+    const client = redis.createClient({
+      url: isDevelopment ? null : process.env.OPENREDIS_URL,
+      retry_strategy: this.handleRedisRetry,
```

Comment:
Rather than using `setTimeout` we can use Redis' built in features: https://github.com/NodeRedis/node-redis#options-object-properties